mindsdb/api/executor/sql_query/steps/fetch_dataframe_partition.py

```diff
@@ -0,0 +1,216 @@
+
+    def concat_results(self, results):
+        df_list = []
+        for res in results:
+            df, col_names = res.to_df_cols()
+            if len(df) > 0:
+                df_list.append(df)
+
+        data = ResultSet()
+        if len(df_list) > 0:
+            data.from_df_cols(pd.concat(df_list), col_names)
+
+        return data
+
+    def exec_sub_steps(self, df):
+        input_data = ResultSet()
+
+        input_data.from_df(
+            df,
+            table_name=self.table_alias[1],
+            table_alias=self.table_alias[2],
+            database=self.table_alias[0]
+        )
+
+        # execute with modified previous results
+        steps_data2 = self.steps_data.copy()
+        steps_data2[self.current_step_num] = input_data
+
+        sub_data = None
+        for substep in self.substeps:
+            sub_data = self.sql_query.execute_step(substep, steps_data=steps_data2)
+            steps_data2[substep.step_num] = sub_data
+        return sub_data
+
+    def fetch_threads(self, run_query, query, thread_count=None):
+        # process batches in threads
+
+        # create communication queues
+        queue_in = queue.Queue()
+        queue_out = queue.Queue()
+        self.stop_event = threading.Event()
+
+        if thread_count is None:
+            thread_count = get_max_thread_count()
+
+        # 3 tasks per worker during 1 batch
+        partition_size = int(run_query.batch_size / thread_count / 3)
+        # min partition size
+        if partition_size < 10:
+            partition_size = 10
+
+        # create N workers pool
+        workers = []
+        for i in range(thread_count):
+            worker = threading.Thread(target=self._worker, daemon=True, args=(ctx.dump(), queue_in,
+                                                                              queue_out, self.stop_event))
+            worker.start()
+            workers.append(worker)
+
+        results = []
+        while True:
+
+            # fetch batch
+            query2 = run_query.get_partition_query(self.current_step_num, query)
+            df, columns_info = self.dn.query(
+                query=query2,
+                session=self.session
+            )
+
+            if df is None or len(df) == 0:
+                # TODO detect circles: data handler ignores condition and output is repeated
+
+                # exit & stop workers
+                self.stop_event.set()
+                break
+
+            max_track_value = run_query.get_max_track_value(df)
+
+            # split into chunks and send to workers
+            sent_chunks = 0
+            for df2 in split_data_frame(df, partition_size):
+                queue_in.put(df2)
+                sent_chunks += 1
+
+            for i in range(sent_chunks):
+                res = queue_out.get()
+                if 'error' in res:
+                    self.close_workers(workers)
+                    raise RuntimeError(res['error'])
+
+                if res['data']:
```

Comment:
should it have the same order as initial data?

Reply:
it is one of the problem of using threads
we can add restriction, something like: raise exception if query has `order by` and thread are used

Reply:
may be send data to the queue with id of the chunk, and assemble in `concat_results` considering chunks numbers?

Reply:
implemented in the latest commit